### PR TITLE
Fix: Using the Latest  GitHub Release type filter should return the release marked as latest

### PR DESCRIPTION
**Pull request by olblak**
Fix #969

Another approach than https://github.com/updatecli/updatecli/pull/988

## Test

To test this pull request, you can run the following commands:

```shell
cp <to_package_directory>
go test
```

### CODE REVIEW

**Comment by dduportal** [comment]
Proposal:

```
query getLatestRelease($owner: String!, $repository: String!) {
  rateLimit {
    cost
    remaining
    resetAt
  }
  repository(owner: $owner, name: $repository) {
    latestRelease {
      id
      name
      tagName
    }
  }
}
```

when "Latest: true" WDYT?

**Comment by olblak** [comment]
that 

> Proposal:
> 
> ```
> query getLatestRelease($owner: String!, $repository: String!) {
>   rateLimit {
>     cost
>     remaining
>     resetAt
>   }
>   repository(owner: $owner, name: $repository) {
>     latestRelease {
>       id
>       name
>       tagName
>     }
>   }
> }
> ```
> 
> when "Latest: true" WDYT?

That would be a better query as it would reduce the amount of information retrieved from the API which ultimately will benefit the api limit usage

File: pkg/plugins/scms/github/releaseType.go [hunk 1]
```diff
@@ -2,13 +2,13 @@ package github
...
+	// "Draft" enable/disable GitHub draft release
 	Draft bool
-	// PreRelease enable/disable GitHub PreRelease
+	// "PreRelease" enable/disable GitHub PreRelease
 	PreRelease bool
-	// Release enable/disable GitHub release
+	// "Release" enable/disable GitHub release
 	Release bool
-	// Latest enable/disable the latest Github Release
+	// "Latest" if set to true will only filter the release flag as latest.
```
Comment (dduportal) [inline pkg/plugins/scms/github/releaseType.go:11]:
```suggestion
	// "Latest" only returns the release marked as "Latest" in GitHub releases
```

File: pkg/plugins/scms/github/releaseType.go [hunk 2]
```diff
@@ -2,13 +2,13 @@ package github
 
 // ReleaseType specifies accepted GitHub Release type
 type ReleaseType struct {
-	// Draft enable/disable GitHub draft release
+	// "Draft" enable/disable GitHub draft release
```
Comment (dduportal) [inline pkg/plugins/scms/github/releaseType.go:5]: Are these comments visible by end suer through JSON schema, or is it development only?

Comment (olblak) [reply pkg/plugins/scms/github/releaseType.go:5]: Yes it's visible to endusers via www.updatecli.io and jsonschema. 
Comments must be one line before the variable

**Comment by dduportal** [comment]
I was able to update the unit test (`pkg/plugins/scms/github/release_test.go`) to cover:

- The issue I see on the main branch
- The "Draft" and "Prerelease" cases from #969 
- The "Latest" case as you described it + implemented it in the current PR

----

**On the main branch (commit 482425f202d7378106dc02b640b2504024010aa4)**

```
ok  	github.com/updatecli/updatecli/pkg/plugins/scms/github	0.684s	coverage: 41.3% of statements
```

Capture d’écran 2022-11-19 à 14 25 34

----

<details>
<summary>Click to view the new content for the `pkg/plugins/scms/github/release_test.go` test file</summary>

```
package github

import (
	"fmt"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestSearchReleases(t *testing.T) {
	tests := []struct {
		name         string
		spec         Spec
		releaseType  ReleaseType
		mockedQuery  *releasesQuery
		mockedError  error
		wantReleases []string
		wantErr      bool
	}{
		{
			name: "Case with a mix of releases (draft, prerelease, release) and default ReleaseType filter",
			spec: Spec{
				Owner:      "updatecli",
				Repository: "updatecli",
				Token:      "superSecretTokenForJoe",
				Username:   "joe",
			},
			mockedQuery: &releasesQuery{
				RateLimit: RateLimit{
					Cost:      1,
					Remaining: 5000,
				},
				Repository: struct {
					Releases repositoryRelease `graphql:"releases(last: 100, before: $before, orderBy: $orderBy)"`
				}{
					Releases: repositoryRelease{
						TotalCount: 4,
						PageInfo: PageInfo{
							HasNextPage: false,
							EndCursor:   "Y3Vyc29yOnYyOpK5MjAyMC0wMi0xOVQyMTozMDoxMyswMTowMM4BYOFb",
						},
						Edges: []releaseEdge{
							{
								Cursor: "Y3Vyc29yOnYyOpK5MjAyMi0wMS0xOFQxOTo0Nzo1MyswMTowMM4Da8vn",
								Node: releaseNode{
									Name:    "v0.18.4 🌈",
									TagName: "v0.18.4",
									IsDraft: true,
								},
							},
							{
								Cursor: "Y3Vyc29yOnYyOpK5MjAyMi0wMS0xOFQxOTo0Nzo1MyswMTowMM4Da8vn",
								Node: releaseNode{
									Name:     "v0.18.3",
									TagName:  "v0.18.3",
									IsLatest: true,
								},
							},
							{
								Cursor: "Y3Vyc29yOnYyOpK5MjAyMS0xMS0xM1QxMjozNzoxMiswMTowMM4DK66W",
								Node: releaseNode{
									Name:         "v1.0.0",
									TagName:      "v1.0.0",
									IsPrerelease: true,
								},
							},
							{
								Cursor: "Y3Vyc29yOnYyOpK5MjAyMC0wMi0xOVQyMTozMDoxMyswMTowMM4BYOFb",
								Node: releaseNode{
									Name:    "v0.0.1",
									TagName: "v0.0.1",
								},
							},
						},
					},
				},
			},
			wantReleases: []string{"v0.0.1", "v0.18.3"},
		},
		{
			name: "Case with a mix of releases (draft, prerelease, release) and a ReleaseType with draft, prerelease and releases",
			spec: Spec{
				Owner:      "updatecli",
				Repository: "updatecli",
				Token:      "superSecretTokenForJoe",
				Username:   "joe",
			},
			releaseType: ReleaseType{
				Draft:      true,
				PreRelease: true,
				Release:    true,
			},
			mockedQuery: &releasesQuery{
				RateLimit: RateLimit{
					Cost:      1,
					Remaining: 5000,
				},
				Repository: struct {
					Releases repositoryRelease `graphql:"releases(last: 100, before: $before, orderBy: $orderBy)"`
				}{
					Releases: repositoryRelease{
						TotalCount: 4,
						PageInfo: PageInfo{
							HasNextPage: false,
							EndCursor:   "Y3Vyc29yOnYyOpK5MjAyMC0wMi0xOVQyMTozMDoxMyswMTowMM4BYOFb",
						},
						Edges: []releaseEdge{
							{
								Cursor: "Y3Vyc29yOnYyOpK5MjAyMi0wMS0xOFQxOTo0Nzo1MyswMTowMM4Da8vn",
								Node: releaseNode{
									Name:    "v0.18.4 🌈",
									TagName: "v0.18.4",
									IsDraft: true,
								},
							},
							{
								Cursor: "Y3Vyc29yOnYyOpK5MjAyMi0wMS0xOFQxOTo0Nzo1MyswMTowMM4Da8vn",
								Node: releaseNode{
									Name:     "v0.18.3",
									TagName:  "v0.18.3",
									IsLatest: true,
								},
							},
							{
								Cursor: "Y3Vyc29yOnYyOpK5MjAyMS0xMS0xM1QxMjozNzoxMiswMTowMM4DK66W",
								Node: releaseNode{
									Name:         "v1.0.0",
									TagName:      "v1.0.0",
									IsPrerelease: true,
								},
							},
							{
								Cursor: "Y3Vyc29yOnYyOpK5MjAyMC0wMi0xOVQyMTozMDoxMyswMTowMM4BYOFb",
								Node: releaseNode{
									Name:    "v0.0.1",
									TagName: "v0.0.1",
								},
							},
						},
					},
				},
			},
			wantReleases: []string{"v0.0.1", "v1.0.0", "v0.18.3", "v0.18.4"},
		},
		{
			name: "Case with a mix of releases (draft, prerelease, release) and a ReleaseType with 'Latest'",
			spec: Spec{
				Owner:      "updatecli",
				Repository: "updatecli",
				Token:      "superSecretTokenForJoe",
				Username:   "joe",
			},
			releaseType: ReleaseType{
				Latest: true,
			},
			mockedQuery: &releasesQuery{
				RateLimit: RateLimit{
					Cost:      1,
					Remaining: 5000,
				},
				Repository: struct {
					Releases repositoryRelease `graphql:"releases(last: 100, before: $before, orderBy: $orderBy)"`
				}{
					Releases: repositoryRelease{
						TotalCount: 4,
						PageInfo: PageInfo{
							HasNextPage: false,
							EndCursor:   "Y3Vyc29yOnYyOpK5MjAyMC0wMi0xOVQyMTozMDoxMyswMTowMM4BYOFb",
						},
						Edges: []releaseEdge{
							{
								Cursor: "Y3Vyc29yOnYyOpK5MjAyMi0wMS0xOFQxOTo0Nzo1MyswMTowMM4Da8vn",
								Node: releaseNode{
									Name:    "v0.18.4 🌈",
									TagName: "v0.18.4",
									IsDraft: true,
								},
							},
							{
								Cursor: "Y3Vyc29yOnYyOpK5MjAyMi0wMS0xOFQxOTo0Nzo1MyswMTowMM4Da8vn",
								Node: releaseNode{
									Name:     "v0.18.3",
									TagName:  "v0.18.3",
									IsLatest: true,
								},
							},
							{
								Cursor: "Y3Vyc29yOnYyOpK5MjAyMS0xMS0xM1QxMjozNzoxMiswMTowMM4DK66W",
								Node: releaseNode{
									Name:         "v1.0.0",
									TagName:      "v1.0.0",
									IsPrerelease: true,
								},
							},
							{
								Cursor: "Y3Vyc29yOnYyOpK5MjAyMC0wMi0xOVQyMTozMDoxMyswMTowMM4BYOFb",
								Node: releaseNode{
									Name:    "v0.0.1",
									TagName: "v0.0.1",
								},
							},
						},
					},
				},
			},
			wantReleases: []string{"v0.18.3"},
		},
		{
			name:        "Case with error returned from github query",
			mockedQuery: &releasesQuery{},
			mockedError: fmt.Errorf("Random error from GitHub API."),
			wantErr:     true,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			require.NotNil(t, tt.mockedQuery)

			sut := Github{
				Spec: tt.spec,
				client: &MockGitHubClient{
					mockedQuery: tt.mockedQuery,
					mockedErr:   tt.mockedError,
				},
			}
			tt.releaseType.Init()

			got, err := sut.SearchReleases(tt.releaseType)

			if tt.wantErr {
				assert.Error(t, err)
				return
			}

			require.NoError(t, err)

			assert.Equal(t, tt.wantReleases, got)
		})
	}
}
```

</details>

----

Result of the test on the main branch (failing as expected because of the bug I described):

```
time="2022-11-19T14:26:45+01:00" level=error msg="\tRandom error from GitHub API."
--- FAIL: TestSearchReleases (0.00s)
    --- FAIL: TestSearchReleases/Case_with_a_mix_of_releases_(draft,_prerelease,_release)_and_default_ReleaseType_filter (0.00s)
        /Users/dadou/workspace/updatecli/updatecli/pkg/plugins/scms/github/release_test.go:238: 
            	Error Trace:	/Users/dadou/workspace/updatecli/updatecli/pkg/plugins/scms/github/release_test.go:238
            	Error:      	Not equal: 
            	            	expected: []string{"v0.0.1", "v0.18.3"}
            	            	actual  : []string{"v0.0.1"}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,4 +1,3 @@
            	            	-([]string) (len=2) {
            	            	- (string) (len=6) "v0.0.1",
            	            	- (string) (len=7) "v0.18.3"
            	            	+([]string) (len=1) {
            	            	+ (string) (len=6) "v0.0.1"
            	            	 }
            	Test:       	TestSearchReleases/Case_with_a_mix_of_releases_(draft,_prerelease,_release)_and_default_ReleaseType_filter
    --- FAIL: TestSearchReleases/Case_with_a_mix_of_releases_(draft,_prerelease,_release)_and_a_ReleaseType_with_draft,_prerelease_and_releases (0.00s)
        /Users/dadou/workspace/updatecli/updatecli/pkg/plugins/scms/github/release_test.go:238: 
            	Error Trace:	/Users/dadou/workspace/updatecli/updatecli/pkg/plugins/scms/github/release_test.go:238
            	Error:      	Not equal: 
            	            	expected: []string{"v0.0.1", "v1.0.0", "v0.18.3", "v0.18.4"}
            	            	actual  : []string{"v0.0.1", "v1.0.0", "v0.18.4"}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,5 +1,4 @@
            	            	-([]string) (len=4) {
            	            	+([]string) (len=3) {
            	            	  (string) (len=6) "v0.0.1",
            	            	  (string) (len=6) "v1.0.0",
            	            	- (string) (len=7) "v0.18.3",
            	            	  (string) (len=7) "v0.18.4"
            	Test:       	TestSearchReleases/Case_with_a_mix_of_releases_(draft,_prerelease,_release)_and_a_ReleaseType_with_draft,_prerelease_and_releases
    --- FAIL: TestSearchReleases/Case_with_a_mix_of_releases_(draft,_prerelease,_release)_and_a_ReleaseType_with_'Latest' (0.00s)
        /Users/dadou/workspace/updatecli/updatecli/pkg/plugins/scms/github/release_test.go:238: 
            	Error Trace:	/Users/dadou/workspace/updatecli/updatecli/pkg/plugins/scms/github/release_test.go:238
            	Error:      	Not equal: 
            	            	expected: []string{"v0.18.3"}
            	            	actual  : []string{"v0.0.1", "v0.18.3"}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1,3 @@
            	            	-([]string) (len=1) {
            	            	+([]string) (len=2) {
            	            	+ (string) (len=6) "v0.0.1",
            	            	  (string) (len=7) "v0.18.3"
            	Test:       	TestSearchReleases/Case_with_a_mix_of_releases_(draft,_prerelease,_release)_and_a_ReleaseType_with_'Latest'
FAIL
FAIL	github.com/updatecli/updatecli/pkg/plugins/scms/github	0.289s
FAIL
```

--- 

**Result on this PR (commit c9647687aba921fa7b65f6a7ea57a481b8676533)**

```
ok  	github.com/updatecli/updatecli/pkg/plugins/scms/github	0.688s	coverage: 42.0% of statements
```
```

**Comment by olblak** [comment]
@dduportal Did you push the updated test somewhere ? Or should I copy paste the code from your example?

**Comment by dduportal** [comment]
> @dduportal Did you push the updated test somewhere ? Or should I copy paste the code from your example?

I did not push anything: you have to copy and paste :)

File: pkg/plugins/scms/github/release_test.go [hunk 3]
```diff
@@ -100,7 +289,7 @@ func TestSearchReleases(t *testing.T) {
 					mockedErr:   tt.mockedError,
 				},
 			}
-			releaseType := ReleaseType{}
+			releaseType := tt.releaseType
 			releaseType.Init()
```
Comment (dduportal) [inline pkg/plugins/scms/github/release_test.go:293]:
```suggestion
			tt.releaseType.Init()
```
No need to copy into another variable

**Comment by olblak** [comment]
@dduportal Thanks for your UT suggestion

**Pull request merged**